homeassistant/components/vicare/__init__.py

```diff
@@ -87,20 +66,27 @@ def setup_vicare_api(hass: HomeAssistant, entry: ConfigEntry) -> None:
     vicare_api = vicare_login(hass, entry.data)
 
     device_config_list = get_supported_devices(vicare_api.devices)
+    if number_of_devices := len(device_config_list) > 1:
+        cache_duration = max(
+            DEFAULT_CACHE_DURATION, DEFAULT_CACHE_DURATION * number_of_devices
+        )
+        _LOGGER.debug(
+            "Found %s devices, adjusting cache duration to %s",
+            number_of_devices,
+            cache_duration,
+        )
+        vicare_api = vicare_login(hass, entry.data, cache_duration)
+        device_config_list = get_supported_devices(vicare_api.devices)
 
     for device in device_config_list:
         _LOGGER.debug(
             "Found device: %s (online: %s)", device.getModel(), str(device.isOnline())
         )
 
-    # Currently we only support a single device
-    device = device_config_list[0]
-    hass.data[DOMAIN][entry.entry_id][VICARE_DEVICE_CONFIG_LIST] = device_config_list
-    hass.data[DOMAIN][entry.entry_id][VICARE_DEVICE_CONFIG] = device
-    hass.data[DOMAIN][entry.entry_id][VICARE_API] = getattr(
-        device,
-        HEATING_TYPE_TO_CREATOR_METHOD[HeatingType(entry.data[CONF_HEATING_TYPE])],
-    )()
+    hass.data[DOMAIN][entry.entry_id][DEVICE_CONFIG_LIST] = [
+        (device_config, get_device(entry, device_config))
+        for device_config in device_config_list
+    ]
```

Comment:
Instead of storing anonymous tuples here, maybe store instances of some dataclass which has config and PyViCareDevice attributed? Also, `DEVICE_CONFIG_LIST` is a bit misleading since it's no longer a list of device configs, but a list of tuples (config, PyViCareDevice).

Reply:
I added a data class and renamed it to device list.